Add inspect view to snippets

laymonage opened the pull request:
<!-- Thanks for contributing to Wagtail! 🎉  Please add a description below, explaining the purpose of this pull request - including the issue number of the issue you're fixing (if applicable). -->

This PR adds a generic `InspectView` that can be used for snippets via `SnippetViewSet`. The implementation is heavily based on ModelAdmin's `InspectView`.

Still a bit unsure of how image and documents should be handled though, as I don't really like how we need to import from `wagtail.images` and `wagtail.documents` within `wagtail.admin`.

I tried using the read-only `FieldPanel` but it doesn't handle image nor documents, so I decided not to use it.

_Please check the following:_

-   [x] Do the tests still pass?[^1]
-   [x] Does the code comply with the style guide?
    -   [x] Run `make lint` from the Wagtail root.
-   [x] For Python changes: Have you added tests to cover the new/fixed behaviour?
-   [x] For new features: Has the documentation been updated accordingly?

**Please describe additional details for testing this change**.

Add `inspect_view_enabled = True` to the `PersonViewSet` in bakerydemo.

[^1]: [Development Testing](https://docs.wagtail.org/en/latest/contributing/developing.html#testing)
[^2]: [Browser and device support](https://docs.wagtail.org/en/latest/contributing/developing.html#browser-and-device-support)
[^3]: [Accessibility Target](https://docs.wagtail.org/en/latest/contributing/developing.html#accessibility-targets)

pySilver commented:
@laymonage fantastic work. I'm in the middle of migration to 5.x and this is what I've been missing. I'm going to use it in a fork and get back here in case of problems or suggestions. 

Thanks for your effort and I hope to see this merged soon!

pySilver commented:
@laymonage I think that the navigation header can be valuable addition here. Otherwise inspection feels kind lost.

![CleanShot 2023-07-05 at 00 23 25@2x](https://github.com/wagtail/wagtail/assets/1113925/4e1e2ed8-d392-4000-92b8-fc60aa8d013d)

pySilver commented:
It might be also smart to add "Inspect" action on the edit view (buttons listed in the footer). So user can easily go there. But that's TBD.

laymonage commented:
Thanks @pySilver

> I think that the navigation header can be valuable addition here. Otherwise inspection feels kind lost.

The breadcrumb is something we'd like to revisit in the future. The implementation template is currently not ideal as it's copy-pasted from the page breadcrumbs. Historically, only the page views had breadcrumbs (and only up to the edit view), but we had to add them to Snippets so that we have room for the side panel buttons. Ideally, if we want to add more breadcrumbs, we probably should add breadcrumbs across the whole admin.

There are still some design decisions needed though, because the page breadcrumbs are related to the page tree, which does not apply to other areas of the admin.

> It might be also smart to add "Inspect" action on the edit view (buttons listed in the footer). So user can easily go there. But that's TBD.

Good idea, but I think that should be in the three-dot menu on the breadcrumbs. We want the footer actions to be related to the form. (On another note: we should probably get rid of the "delete" button there, as it seems to be the case for pages)

lb- commented:
To add to the above discussion, there are two issues created to deal with the Breadcrumbs assumptions that they will be for page structures only. 

- https://github.com/wagtail/wagtail/issues/8767
- https://github.com/wagtail/wagtail/issues/8037
- https://github.com/wagtail/wagtail/issues/10118 (Stimulus migration, may need to consider the above two cases)

laymonage commented on wagtail/snippets/views/snippets.py:
```diff
@@ -493,6 +493,11 @@ def get_queryset(self):
 
+class InspectView(generic.InspectView):
+    view_name = "inspect"
+    any_permission_required = ["add", "change", "delete"]
```
Would be nice to reuse Django's `view` permission here, but there will be some other areas that need to be changed, e.g. the permission check for the model index view that uses this function:

https://github.com/wagtail/wagtail/blob/853aad4305440336e2ce6b70769fdf827fe7922f/wagtail/snippets/permissions.py#L13-L19

I decided to leave this for now to avoid making the PR scope too big.

gasman reviewed (changes requested):
Looking good!

To avoid the imports of wagtail.images and wagtail.docs (as well as opening up the possibility of third-party packages like wagtailmedia extending this), I'd suggest setting up an instance of [wagtail.utils.registry.ModelFieldRegistry](https://github.com/wagtail/wagtail/blob/dc747d710693a1be124afe1c18da0fcbcfd478df/wagtail/utils/registry.py#L46) so that you can look up the field and get back an object that knows how to format it. wagtail.images can then set up a handler for foreign keys to Image (just as it does [to register AdminImageChooser as the widget for that field](https://github.com/wagtail/wagtail/blob/dc747d710693a1be124afe1c18da0fcbcfd478df/wagtail/images/apps.py#L24-L27)), wagtail.documents can do the same for Documents, and the fallback handler (registered against the base Field class) will just output the value directly.

pySilver commented:
@gasman or we can leave a simple URL to the media file because this inspect page will always be kinda dead simple and ugly by default. Some table of key/value can be fine.

laymonage commented:
@gasman I've changed the approach to use a generic "field display registry" (in 93360b5) that maps model fields to `Component` classes. I was thinking this might be useful in other places in the future (not just the inspect view), e.g. read-only field panels or the listing views.

I made a separate commit for the base `Field` class in 6c9af56 because it was not necessary for the inspect view to work, but I figured it's still nice to have it anyway.

laymonage commented on wagtail/admin/views/generic/models.py:
```diff
@@ -715,6 +721,114 @@ def get_context_data(self, **kwargs):
+        context = {
+            "label": self.get_field_label(field_name, field),
+            "value": self.get_field_display_value(field_name, field),
+            "component": None,
+        }
+        if isinstance(context["value"], Component):
+            context["component"] = context["value"]
```
Not very happy with this, but the value returned by `get_field_display_value` may not always be a `Component` class as it also handles arbitrary attributes on the model (not just model fields).

I suppose the nice thing about this is that it also lets you return a `Component` instance from a model method.

The main reason why I needed this was so that the template knows when to use `{% component %}`. If you have better ideas let me know!

laymonage commented:
@gasman apparently calling `render_html()` with `{% component %}` will make Django skip any [format localisation](https://docs.djangoproject.com/en/4.2/topics/i18n/formatting/) of the returned value. So, even if the `BaseFieldDisplay` component returns the value directly (without doing e.g. `str()`), the rendered value will be as if you do a `str(value)`. This means things like dates will be rendered as e.g. `2023-13-07`.

If `render_html()` is called from the view, and then in the template we do `{{ rendered_value }}`, then the format localisation is not skipped, and dates will be rendered as e.g. `13 Jul 2023`. However, I don't think this is how we want the `Component` class to be used. (In addition, we also need to pass the `parent_context` manually, which isn't very easy if you don't have reference to the full context where the `render_html()` is used.)

This is why the tests are failing after I added 57c39f0f6570ac04bf975f4bc58ce56b59aff26a. Without it, the value is formatted with the locale and tests will pass. Let me know whether we should keep `BaseFieldDisplay` and update the test, or if you have a better approach. Thanks!

gasman commented:
@laymonage From a quick dig into the Django template engine, it looks like calling [`render_value_in_context`](https://github.com/django/django/blob/02966a30dd31d2b9d35f8c481a448b9bf377895e/django/template/base.py#L1032) from `render_html` (and passing `parent_context` as the context) would have the same effect as rendering a template consisting of `{{ value }}`.

Having said that, if everything works fine without BaseFieldDisplay, then I'd be inclined to avoid overengineering things, and just drop 57c39f0f6570ac04bf975f4bc58ce56b59aff26a entirely.

laymonage commented:
> @laymonage From a quick dig into the Django template engine, it looks like calling [`render_value_in_context`](https://github.com/django/django/blob/02966a30dd31d2b9d35f8c481a448b9bf377895e/django/template/base.py#L1032) from `render_html` (and passing `parent_context` as the context) would have the same effect as rendering a template consisting of `{{ value }}`.
> 
> Having said that, if everything works fine without BaseFieldDisplay, then I'd be inclined to avoid overengineering things, and just drop [57c39f0](https://github.com/wagtail/wagtail/commit/57c39f0f6570ac04bf975f4bc58ce56b59aff26a) entirely.

Nice find! TIL about that function.

I’ve dropped the commit and split the previous one into two: one that implements the generic field display registry, and the one after just replaces the inspect view implementation to use it.

I decided to add a `BaseFieldDisplay` (not registered) to provide a unified interface for these display classes.

gasman reviewed (approved):
LGTM!

jeromelebleu commented:
I have just discovered this new helpful generic view, thanks! I am wondering why did you not use `[django.views.generic.detail.DetailView](https://docs.djangoproject.com/en/4.2/ref/class-based-views/generic-display/#detailview)` as the base class? It would provide the default methods to retrieve the object such as `get_queryset()` which could be really useful - e.g. to select related fields. If using this class instead is welcome, I can work on a PR.

laymonage commented:
> I have just discovered this new helpful generic view, thanks!
> 
> I am wondering why did you not use [`django.views.generic.detail.DetailView`](https://docs.djangoproject.com/en/4.2/ref/class-based-views/generic-display/#detailview) as the base class? It would provide the default methods to retrieve the object such as `get_queryset()` which could be really useful - e.g. to select related fields. If using this class instead is welcome, I can work on a PR.

Glad you find it useful!

I don't think there was a strong reason, but Django's class-based views are [notoriously complex](https://ccbv.co.uk) so I opted to keep things simple by extending from `TemplateView` instead. It makes it clear that we're only using the template-rendering mechanism from the Django base class and the logic you need to know is pretty much all laid out in the view class directly.

One thing I don't really like about Django's `BaseDetailView`/`BaseUpdateView`/`BaseCreateView` is that `get_object()` called in `get()` and `post()` instead of `setup()`:
- https://github.com/django/django/blob/e41f9f94505f690525f16af8bd10e62e22bc0207/django/views/generic/detail.py#L107-L108
- https://github.com/django/django/blob/e41f9f94505f690525f16af8bd10e62e22bc0207/django/views/generic/edit.py#L200-L206

Looking at their code, the duplicated overrides for `get()` and `post()` could've been simplified if they had moved `self.object = self.get_object()` to `setup()` and it would also guarantee that all methods have access to `self.object` in the same way. Of course, this doesn't matter much in the `InspectView` as you'd only use `GET`.

Anyway, that was just a rant 👀 

> It would provide the default methods to retrieve the object such as get_queryset() which could be really useful - e.g. to select related fields.
Can't this be done by overriding `get_object()`? Is there other use cases where this can't be accomplished by overriding `get_object()` and where Django's `DetailView` comes in handy?

jeromelebleu commented:
I agree with the complexity and the changes which could be done to these class-based views. Moving `self.object = self.get_object()` to `setup()` could also ease the permission checking for an instance - but that's another point!

> Can't this be done by overriding get_object()? Is there other use cases where this can't be accomplished by overriding get_object() and where Django's DetailView comes in handy?

It is less convenient here since you have to rewrite this method - e.g.:
```python
def get_object(self):
    return get_object_or_404(self.model.objects.all().select_related('some_field'), pk=unquote(self.pk))
```
… instead of just "extending" the base QuerySet like this:
```python
def get_queryset(self):
    return super().get_queryset().select_related('some_field')
```

> I wouldn't really mind updating this to use BaseDetailView if the benefit is clear and worth the added complexity, but if we start using it we should probably use it in other similar views too e.g. UnpublishView, RevisionsUnscheduleView, etc.

I don't mind having `BaseDetailView` too honestly, I asked to understand your thoughts. I would just suggest - like you - to be consistent in all these `admin.views.generic.models` views and always provide the same mechanism to retrieve the object - i.e. `get_object()` and `get_queryset()` - either by using Django's class-based views or by writing our own logic.
